source/slang/slang-ir-eliminate-phis.cpp

```diff
@@ -0,0 +1,894 @@
+        // parameter is.
+        //
+        Index paramIndex = kInvalidIndex;
+
+        // When there is a cyclic dependency between arguments at
+        // a branch site, no sequence of plain assignments (without
+        // additional copies) will suffice.
+        //
+        // When we end up having to break cycles, we do so by loading
+        // a copy of the value in one of the per-parameter temporaries.
+        // Any subsequent branch arguments that referenced the parameter
+        // will need to use that copy instead.
+        //
+        // In order to make sure that we properly reference the loaded
+        // copy instead of the original argument in such cases, we use
+        // a pointer field for each argument that points to a location
+        // where the up-to-date value can be found.
+        //
+        // For most arguments this will always just point to `originalVal`,
+        // but for arguments that refer to a block parameter, this will
+        // point to the `currentVal` field of the corresponding `ParamInfo`.
+        //
+        IRInst** currentValPtr = nullptr;
+    };
+    enum { kInvalidIndex = -1 };
+
+    // A lot of the logic in this pass is concerned with the process of
+    // emitting *assignments* from branch arguments to block parameters.
+    // Those assignments are implicit in our SSA IR, but this pass needs
+    // to make them explicit.
+    //
+    // In order to emit assignments in an order that minimizes the number
+    // of temporaries/copies that appear in output code, we need a way
+    // to track which assignments have been done, which are ready to be done,
+    // and which are "blocked" for some reason. We thus associate each
+    // assignment with an integer _state_ which is in one of three cases:
+    //
+    // * _done_ (`-1`): any instructions needed for the assignment have been emitted
+    //
+    // * _ready_ (`0`): the assignment can be emitted without causing any problems
+    // 
+    // * _blocked_ (`N > 0`): the assignment cannot be emitted yet, because there
+    // are `N` other not-yet-done assignments that need to read the value of the
+    // parameter that this assignment wants to write. The reads need to be able
+    // to proceed before the write can go through.
+    //
+    enum
+    {
+        kState_Done = -1,
+        kState_Ready = 0,
+    };
+
+    // There is a one-to-one correspondance between:
+    //
+    // * The phis/parameters of a particular `block`
+    // * The arguments passed at some branch to `block`
+    // * The assignments that need to be performed at that branch site
+    //
+    // We thus use a single structure to track all of that information,
+    // which is handy but also requires careful thought at use sites
+    // about which version of the information is relevant.
+    //
+    struct PhiInfo
+    {
+        ParamInfo param;
+        ArgInfo arg;
+        Count state = kState_Ready;
+    };
+    List<PhiInfo> phiInfos;
```

Comment:
It might be better to make these member variables prefixed. Doing so would make clear at usage sites that they belong to the context. If the class could be accessed externally, and they were publicly accessible (ie 'they are part of the interface') then that might override this consideration.

Reply:
Yeah, I bounce back and forth on this stuff, and I know this particular type is actually inconsistent (some stuff using `m_` prefix and some not). My experience is that for context types like this I often go through multiple refactoring steps where things evolve between local variable <-> function parameter <-> member variable of context type, and it is frustrating if exactly one of those cases requires touching up all the use sites.

If C++ supported local functions more nicely this all might be moot.